src/dotnet/devcontainer-feature.json

```diff
@@ -1,35 +1,31 @@
 {
     "id": "dotnet",
-    "version": "1.1.4",
+    "version": "2.0.0",
     "name": "Dotnet CLI",
     "documentationURL": "https://github.com/devcontainers/features/tree/main/src/dotnet",
-    "description": "Installs the .NET CLI. Provides option of installing sdk or runtime, and option of versions to install. Uses latest version of .NET sdk as defaults to install.",
+    "description": "Installs .NET and its CLI. Offers options to install multiple SDK versions.",
     "options": {
         "version": {
             "type": "string",
             "proposals": [
                 "latest",
-                "7",
-                "6",
-                "3.1"
+                "lts",
+                "8.0",
+                "7.0",
+                "6.0"
             ],
             "default": "latest",
-            "description": "Select or enter a dotnet CLI version. (Available versions may vary by Linux distribution.)"
+            "description": "Version of .NET SDK to install. Use 'latest' for the latest version, 'lts' for the latest LTS version, 'X.Y' for a specific release, or 'X.Y.Z' for a specific version."
         },
-        "runtimeOnly": {
```

Comment:
My 2¢ is that it would be better to offer the runtimes as separate features, as that would solve my concerns with offering it as a `runtimeOnly` boolean option.

- Separate features allow users to install a **precise** mix of SDKs and Runtimes
- It will be easier to handle differences between the SDK versioning and Runtime versioning
    - Eliminate the need for `if [ "${DOTNET_RUNTIME_ONLY}" = "true" ]; then ...` conditional code
    - [How .NET is versioned](https://learn.microsoft.com/en-us/dotnet/core/versions/)


``` json
{
    "features": {
        "ghcr.io/devcontainers/features/dotnet:2": { "version": "latest", "additionalVersions": [] },
        "ghcr.io/devcontainers/features/dotnet-runtime:1": { "version": "latest", "additionalVersions": [] },
        "ghcr.io/devcontainers/features/aspnetcore-runtime:1": { "version": "latest", "additionalVersions": [] },
    }
}
```

Reply:
I agree with @sliekens that the runtime is best suited for its own feature. I recognize that both features would use this `dotnet-install.sh` script and this overall structure, but the runtime is certainly a distinct tool than the SDK.

This reminds me of the situation with our Python and Jupyter features @samruddhikhandale. We discontinued the Jupyter feature in favor of an option on the Python feature. In hindsight, I think most of us agree that was a confusing decision.

Reply:
Discussed this with the team, historically `runtimeOnly` option is used by folks who simply need to run .NET tools/programs and who uses other programming languages for development. Installing the runtime can save some disk space compared to installing the SDKs. 

I like the idea of isolating it to a new Feature, however, if `runtimeOnly` Feature opiton's usage is pretty low, not sure if we should invest time in creating or maintaining this new Feature. 

Looping in @bderusha who originally added this Feature option, any thoughts about removing this option from the the Feature? or would appreciate your insights on it's usage/use cases. Thank you!

Reply:
I can't speak to how the maintainers should weigh the user value VS maintenance cost of adding dotnetcore and ASPcore runtime features, which I agree is the best user experience option.

Keeping the old boolean is not ideal, but it is not the only alternative. IMHO it would be straight forward to offer support for a version prefix like `dotnetcore#3.1`.  Split on the special character, normalize it, and pass the value to via the `--runtime` parameter and let the install script manage the rest. This might strike a better balance between maintenance and feature support.

It would be great if you could pass the raw prefix to `--runtime`, but the magic values would need to be `aspnetcore` and `dotnet`. The latter being not descriptive enough for me, personally.

Reply:
Or maybe something like this? Naming was never my strong suit but you get the picture.

``` json
{
    "features": {
        "ghcr.io/devcontainers/features/dotnet:2": {
            "version": "latest",
            "additionalVersions": "lts",
            "dotnetRuntimeVersions": "3.1,5.0",
            "aspnetcoreRuntimeVersions": "3.1,5.0"
        }
    }
}
```

Reply:
Yes, I like the explicit nature of this rather than relying on documented magic strings. Looking at your example I am wondering if it might be cleanest to just have three options:

```
{
    "features": {
        "ghcr.io/devcontainers/features/dotnet:2": {
            "dotnetSdkVersions": "lts, 5.0",
            "dotnetcoreRuntimeVersions": "3.1",
            "aspnetcoreRuntimeVersions": "latest"
        }
    }
}
```
And signal the most common case by making `dotnetSdkVersions` default to `latest` and the rest default to empty string.

Reply:
It's my expectation that your example above would **not** disable installation of the SDK. If `dotnetSdkVersions` defaults to `latest` and it is left unspecified (per your example) the resulting environment variables would be:

```
DOTNETSDKVERSIONS="latest"
DOTNETCOREVERSIONS=""
ASPNETCOREVERSIONS="3.1"
```

A user wanting to install only that single runtime would supply:
```
{
    "features": {
        "ghcr.io/devcontainers/features/dotnet:2": {
             "dotnetSdkVersions": "",
             "aspnetcoreRuntimeVersions": "3.1"
        }
    }
}
```
That feels pretty intuitive so long as that default is documented.

Reply:
Is it possible for the install script to distinguish missing config from a configured empty string?

Reply:
Check out the docs for how dev containers handles [options resolution](https://containers.dev/implementors/features/#option-resolution-example). TL;DR: The install script doesn't distinguish between empty VS missing, but the tool that parses the feature object and generated the ENV VARs does and will apply the configured defaults for missing options.

Reply:
Ah, so actually this code at the top is not good? As it would override a configured empty string.

``` sh
DOTNET_VERSION="${VERSION:-"latest"}"
```

Reply:
Removed in 2b2e683da219ca23b892c49376b0449aeb38af84

Now I'm waiting for a decision on what to do with runtime

1. Introduce separate features for each runtime
    - offers the best UX
    - ⚠️ bigger maintenance cost
2. Introduce new options to configure runtime versions
    - also acceptable UX
    - reduced maintenance cost compared to separate features
3. Keep the `runtimeOnly` option as it was before
    - ⚠️ no mixing of SDK and Runtimes possible
    - ⚠️ no ability to specify `dotnet` or `aspnetcore` runtime 
    - ⚠️ more complicated input validation code
4. Discontinue runtime only installations
    - the SDK includes all runtimes anyway
    - ⚠️ bigger image size might not be justifiable if you only need the runtime

I think 2) is clearly ahead in the race so I would like to propose this configuration schema.

``` js
{
    "version": "latest",
    "runtime": undefined | 'dotnet' | 'aspnetcore', // undefined for SDK install (default)
    "additionalSdkVersions": "5.0, 6.0, 7.0",
    "additionalDotnetRuntimeVersions": "5.0, 6.0, 7.0",
    "additionalAspNetCoreRuntimeVersions": "5.0, 6.0, 7.0",
};
```

Reply:
Is there a reason to prefer the idea of "additional" versions to just using a single list of versions (for each type)?

```
{
    "sdkVersions": "5.0, 6.0, 7.0",  // default to "latest"
    "dotnetRuntimeVersions": "5.0, 6.0, 7.0", // default to ""
    "aspNetCoreRuntimeVersions": "5.0, 6.0, 7.0", // default to ""
};
```

Reply:
I'd keep `version` because it is special cased by the CLI to make shorthand configuration possible.

![image](https://github.com/devcontainers/features/assets/1583241/e8f79b38-71f3-4943-97cb-4c63c027be3d)
https://containers.dev/implementors/features/#devcontainer-json-properties

Special treatment in the CLI:

https://github.com/devcontainers/cli/blob/413c5f790f3a4972bf6f05b71d71ac946c87bc7c/src/spec-configuration/containerFeaturesConfiguration.ts#L1303C3-L1305

Reply:
Ah yes, keeping that shorthand is nice. Perhaps an approach that could work to keep `version` but eliminate the need for `kind` is to take a similar approach here as is done for java:

That is, allowing a version to be `"none"` (See java example [here](https://github.com/devcontainers/images/blob/main/src/java/.devcontainer/devcontainer.json#L7-L9))

The use case for Java is that the feature allows you to easily add other binaries like Maven and Gradle, but if you are starting from the base Java devcontainer image you don't want the feature to reinstall java. Since we treat `version` as a shorthand for `SDKversion`, `"none"` is a reasonable choice of SDK for anyone wanting to install standalone runtimes.

Reply:
Perfect! I think consistency in configuration between features is very good.

Updated proposal for runtime-only configuration:

``` js
{
    "version": "none",
    "dotnetRuntimeVersions": "5.0, 6.0, 7.0",
    "aspNetCoreRuntimeVersions": "5.0, 6.0, 7.0"
}
```

SDK only

``` js
{
    "version": "latest",
    "additionalVersions": "lts"
}
```

Mix

``` js
{
    "version": "latest",
    "additionalVersions": "lts",
    "dotnetRuntimeVersions": "5.0",
    "aspNetCoreRuntimeVersions": "5.0"
}
```

(I'm not sure about empty string for `version` - assume `latest` or `none`?)

Reply:
My preference would be to treat empty string the same as `none`, but there's a reasonable argument to be made for either. As long as it's documented I would be supportive of whichever you choose.

Reply:
The proposal in https://github.com/devcontainers/features/pull/628/files#r1296984833 looks amazing! 🎉  Thank you! 

> (I'm not sure about empty string for version - assume latest or none?)

I think `latest` makes the most sense to be in sync with other Features.

Reply:
@bderusha @samruddhikhandale this is now implemented and documented

I tried to organize helper functions into separate files, but I'm not sure about the approach. The files are sourced in `install.sh` where I call their functions. There is a little bit of shared state that is not passed as an argument. (`DOTNET_INSTALL_SCRIPT` is defined in `install.sh` but only used in the helper functions.) I'm not sure how to make that less ugly or if I should just pass everything as function arguments.

Reply:
It is not clear to me what the custom scripts are doing that isn't already handled by `vendor/dotnet-install.sh`. I think that the needs of this feature work well with the defaults to allow you to just pass the version directly into `vendor/dotnet-install.sh` for all cases except `lts`, which can be handled.

```sh
for version in "${versions[@]}"; do
  if [ "${version]" == "lts" ]; then
    args="--version latest --channel LTS --install-dir $DOTNET_INSTALL_DIR --no-path"
  else
    args="--version ${version} --install-dir $DOTNET_INSTALL_DIR --no-path"
  fi

  "$DOTNET_INSTALL_SCRIPT" "${args}"
done

## Repeat for dotnetRuntimeVersions and aspNetCoreRuntimeVersions with appropriate --runtime arg as well
```

Would something like this work or did I miss something? It certainly reduces the footprint

Reply:
@bderusha to clarify, the vendor script usage is really not as simple as you (and I) expected:

* `--version` is only used for an exact version like `6.0.412` 
* `--channel` is used for everything else (and optionally a `--quality` but I chose not to implement it)

So, a lot of the complexity comes from deciding when to use `--channel` instead of `--version`

Examples

* `--channel LTS`
* `--channel STS` (not implemented, we offer `latest` instead) (I don't think STS means latest, needs confirmation)
* `--channel 6.0` for pinning major.minor version
* `--channel 6.0.1xx` for pinning major.minor version and feature band 1 (SDK only)

More complexity comes from figuring out which channel contains the latest version, STS or LTS. This complexity could be removed if STS does actually always point to the latest version, but I have not found any indication of that being true.

In the end, I chose to handle the cases when I know for sure you need to use the `--channel` parameter, and only use `--version` as a fallback.

Reply:
> I tried to organize helper functions into separate files, but I'm not sure about the approach. The files are sourced in install.sh where I call their functions. There is a little bit of shared state that is not passed as an argument. (DOTNET_INSTALL_SCRIPT is defined in install.sh but only used in the helper functions.) I'm not sure how to make that less ugly or if I should just pass everything as function arguments.

Looking at all three files (`install-aspnetcore-runtime.sh`, `install-dotnet-runtime.sh`, `install-dotnet-sdk.sh`), seems like the functions look way too similar, and seems redundant. 

Suggestion: Adding a new file instead of adding functions to the `install.sh` is fine and acceptable. However, can we have one file instead of three?
- eg

```
fetch_latest_aspnetcore_runtime_version_in_channel() {
    local channel="$1"
    wget -qO- "https://dotnetcli.azureedge.net/dotnet/aspnetcore/Runtime/$channel/latest.version"
}
```

Can be refactored like 👇 

```
fetch_latest_version_in_channel() {
    local channel="$1"
    local type="$2"
    wget -qO- "https://dotnetcli.azureedge.net/dotnet/$type/$channel/latest.version"
}
```

Similarly, can we update other two functions as well?

Reply:
Absolutely, taking care of it

Reply:
@samruddhikhandale check again, I think you will be pleased with the result